src/Storages/ObjectStorage/DataLakes/DeltaLake/DeltaLakeSink.cpp

```diff
@@ -17,49 +24,77 @@ DeltaLakeSink::DeltaLakeSink(
     ContextPtr context_,
     SharedHeader sample_block_,
     const std::optional<FormatSettings> & format_settings_)
-    : StorageObjectStorageSink(
-        DeltaLake::generateWritePath(
-            delta_transaction_->getDataPath(),
-            configuration_->format),
-        object_storage_,
-        configuration_,
-        format_settings_,
-        sample_block_,
-        context_)
+    : SinkToStorage(sample_block_)
+    , WithContext(context_)
     , delta_transaction(delta_transaction_)
     , object_storage(object_storage_)
+    , configuration(configuration_)
+    , format_settings(format_settings_)
+    , sample_block(sample_block_)
+    , data_file_max_rows(context_->getSettingsRef()[Setting::delta_lake_insert_max_rows_in_data_file])
+    , data_file_max_bytes(context_->getSettingsRef()[Setting::delta_lake_insert_max_bytes_in_data_file])
 {
     delta_transaction->validateSchema(getHeader());
 }
 
+DeltaLakeSink::StorageSinkPtr DeltaLakeSink::createStorageSink() const
+{
+    return std::make_unique<StorageObjectStorageSink>(
+        DeltaLake::generateWritePath(
+            delta_transaction->getDataPath(),
+            configuration->format),
+        object_storage,
+        configuration,
+        format_settings,
+        sample_block,
+        getContext());
+}
+
 void DeltaLakeSink::consume(Chunk & chunk)
 {
     if (isCancelled())
         return;
 
-    written_bytes += chunk.bytes();
-    StorageObjectStorageSink::consume(chunk);
+    if (data_files.empty()
+        || data_files.back().written_bytes >= data_file_max_bytes
+        || data_files.back().written_rows >= data_file_max_rows)
+    {
+        data_files.emplace_back(createStorageSink());
+    }
+
+    auto & data_file = data_files.back();
+    data_file.written_bytes += chunk.bytes();
```

Comment:
I had an issue where `chunk.bytes()` was much smaller than the actual parquet file size, which caused Spark to skip some Iceberg data files (because the file size in the metadata was too small, so it assumed there was no data file). I suspect the same issue could occur here as well, so it’s better to calculate the size using `buffer->count().`